download-utils/src/lib.rs

```diff
@@ -23,11 +23,41 @@ fn new_spinner_progress_bar() -> ProgressBar {
     progress_bar
 }
 
-pub fn download_file(
+/// Structure modeling information about download progress
+#[derive(Debug)]
+pub struct DownloadProgressRecord {
+    // Duration since the beginning of the download
+    pub elapsed_time: Duration,
+    // Duration since the the last notification
+    pub last_elapsed_time: Duration,
+    // the bytes/sec speed measured for the last notification period
+    pub last_throughput: f32,
+    // the bytes/sec speed measured from the beginning
+    pub total_throughput: f32,
+    // total bytes of the download
+    pub total_bytes: usize,
+    // bytes downloaded so far
+    pub current_bytes: usize,
+    // percentage downloaded
+    pub percentage_done: f32,
+    // Estimated remaining time (in seconds) to finish the download if it keeps at the the last download speed
+    pub estimated_remaining_time: f32,
+    // The times of the progress is being notified, it starts from 1 and increments by 1 each time
+    pub notification_count: u64,
+}
+
+/// This callback allows the caller to get notified of the download progress modelled by DownloadProgressRecord
+/// Return "true" to continue the download
+/// Return "false" to abort the download
+pub fn download_file<F>(
     url: &str,
     destination_file: &Path,
     use_progress_bar: bool,
-) -> Result<(), String> {
+    progress_notify_callback: &Option<F>,
+) -> Result<(), String>
+where
+    F: Fn(&DownloadProgressRecord) -> bool,
```

Comment:
nit: `Fn(&DownloadProgressRecord) -> bool` is repeated multiple times.  You could wrap that in a type and then document the true/value return behaviour at the type definition

Reply:
Let me see to that

Reply:
Actually I attempted a type and I have run into all sort of issues with the type alias. Have not found a way to appease putting it into the Option, and passing around to the DownloadProgress struct. I am still struggling a bit with rust... I keep looking but I'd like to address that in a separate PR to reduce churns  with review and handling merge conflicts.